src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -687,4 +753,102 @@ public static string GetTempPath(string workingDir)
             return Environment.GetEnvironmentVariable("SYSTEMROOT");
         }
     }
+
+    internal class FileMutex : IDisposable 
+    {
+        public readonly bool OwnsMutex;
+        public readonly FileStream Stream;
+        public readonly string FilePath;
+
+        public bool Locked { get; private set; }
+
+        public FileMutex(string name) 
+        {
+            FilePath = Path.Combine(Path.GetTempPath(), name);
+            try 
+            {
+                Stream = new FileStream(FilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = false;
+            } 
+            catch (FileNotFoundException) 
+            {
+                Stream = new FileStream(FilePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = true;
+            }
+        }
+
+        public bool Lock()
+        {
+            if (Locked)
+                return true;
+
+            try 
+            {
+                Stream.Lock(0, 0);
+                Locked = true;
+                return true;
+            } 
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        public void Unlock()
+        {
+            if (!Locked)
+                return;
+            Stream.Unlock(0, 0);
+            Locked = false;
+        }
+
+        public void Dispose()
+        {
+            if (Locked)
+                Unlock();
+            Stream.Dispose();
+            // FIXME: We need to ensure we don't delete the lock while someone else holds it
+            if (OwnsMutex)
+                File.Delete(FilePath);
```

Comment:
The `File.Delete` API can throw under normal circumstances (think a virus checker locking a file). That can't take down the server or client process. Need to guard this operation with a `try / catch`.